rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+
+Serving a proper Kibana application on the root route of the preboot HTTP server implies that we'll also have a chance to replace the static `Kibana server is not ready yet` string with a more helpful and user-friendly application.
+
+### 3.2.9 Preboot service
+
+To support interactive applications at the `preboot` stage we should allow preboot plugins to pause Kibana startup sequence. This functionality will be exposed by the new Preboot service, and will be available to the preboot plugins via `CorePrebootSetup`. Preboot plugins will be able to provide a promise to hold a boot sequence for as much as needed, and also let Kibana know if it has to reload configuration before it enters the `setup` stage.
+
+```ts
+export interface PrebootServiceSetup {
+  readonly isBootOnHold: () => boolean;
+  readonly holdBootUntilResolved: (
+    promise: Promise<{ shouldReloadConfig: boolean } | undefined>
+  ) => void;
+}
+```
+
+Internal Preboot service contract will also expose `waitUntilCanBoot` method that bootstrap process can use to know when it can proceed to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some existing properties have been omitted for brevity.
+export interface InternalPrebootServiceSetup {
+  readonly waitUntilCanBoot: () => Promise<{ shouldReloadConfig: boolean }>;
+}
+```
+
+### 3.2.10 Bootstrap
+
+We'll update Kibana bootstrap sequence to include `preboot` stage and to conditionally reload configuration before proceeding to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some functionality and checks have been omitted for brevity.
+const { preboot } = await root.preboot();
+
+const { shouldReloadConfig } = await preboot.waitUntilCanBoot();
+if (shouldReloadConfig) {
+  await reloadConfiguration('pre-boot request');
+}
+
+await root.setup();
+await root.start();
+```
+
+It's not yet clear if we need to adjust the base path proxy to account for this new lifecycle stage (see [unresolved question 3](#83-development-mode-and-basepath-proxy)).
+
+# 4. Drawbacks
+
+The main drawback is that proposed changes affect quite a few Kibana Core services that may impose a risk of breaking something in the critical parts of Kibana.
+
+# 5. Alternatives
+
+The most viable alternative to support interactive setup mode for Kibana was a standalone application that would be completely separated from Kibana. We ruled out this option since we won't be able to leverage existing and battle-tested Core services, UI components, and development tools. This would make the long-term maintenance burden unreasonably high.
```

Comment:
Another alternative could be to avoid blocking Kibana start until the connection to Elasticsearch is established. @pgayvallet would it be covered by https://github.com/elastic/kibana/issues/96626? Of course, it's not something we'll add tomorrow, but if we think it's the right approach, we can temporarily limit ourselves to any quick hack-ish solution.